### PR TITLE
2259 error message - "auto-call" error

### PR DESCRIPTION
**Pull request by bburdette**
revised this "auto-call" error which I've found to be somewhat inscrutable in the past.  This is intended to address (in part) issue #2259.

- UndefinedVarError seemed more appropriate than TypeError
- added Pos
- big error message with examples.
- link to a nix pill for further explanation.

Current output is like this:

image

Revised error output is like so:
 
image

This reflect my perhaps imperfect understanding of autoCall, feedback on that is welcome.

### CODE REVIEW

**Comment by bburdette** [comment]
Some slight changes:

image

**Comment by edolstra** [comment]
I think very long error messages should be avoided since then we just end up duplicating the manual. The error message can also get out of date, for instance it talks about `nix-build` and a specific (non-flake) style of importing Nixpkgs. It may also lead the user astray that we're talking about `callPackage` but the error message might have nothing to do with `callPackage`.

 It's probably better to have a shorter message that links to the manual for more information (*not* Nix pills which are not an authoritative reference for Nix).

File: src/libexpr/eval.cc [hunk 1]
```diff
@@ -1370,7 +1365,31 @@ void EvalState::autoCallFunction(Bindings & args, Value & fun, Value & res)
             if (j != args.end()) {
                 actualArgs->attrs->push_back(*j);
             } else if (!i.def) {
-                throwTypeError("cannot auto-call a function that has an argument without a default value ('%1%')", i.name);
+                throwUndefinedVarError(i.pos, R"(cannot auto-call a function that has an argument without a default value ('%1%')
+
+An 'auto-call' is when a nix expression is evaluated without any external arguments.
```
Comment (edolstra) [inline src/libexpr/eval.cc:1370]: This is not correct. An auto-call is when a top-level function is called *with* external arguments *or* with default values if no matching argument is provided.

It's probably better to say something like `you need to pass "--arg %1% ...'  or '--argstr %1% ...'".

Comment (edolstra) [reply src/libexpr/eval.cc:1370]: `UndefinedVarError` is probably misleading. There is no undefined variable, just a missing argument.

**Comment by bburdette** [comment]
Thanks for the feedback!

- pared down the error message
- error is more generic rather than callPackage centric
- docs link goes to function reference in the manual.
- added a MissingArgumentError rather than using TypeError or UndefinedVarError.

Here's the result:

image

**Comment by bburdette** [comment]
Thanks for the merge!

**Pull request merged**